### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by Xelsoid**
image

### CODE REVIEW

File: src/__tests__/helpers.test.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      ({ year, country, expected }) => {
+        const result = validateInput({ year, country });
+        expect(result).toEqual(expected);
+      },
+    );
+
+    it.each([{ year: 2024 }, { year: 1 }, { year: 20233 }, { year: -102 }])(
+      "should throw error in case not valid year: $year passed",
+      ({ year }) => {
+        expect(() => validateInput({ year, country: "GB" })).toThrow(
```
Comment (ShaxxKh) [inline src/__tests__/helpers.test.ts:22]: another way to test error handling
```suggestion
        await expect(validateInput({ year, country: "GB" })).rejects.toThrow(
```

File: src/__tests__/helpers.test.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      ({ country }) => {
+        expect(() => validateInput({ year: 2023, country })).toThrow(
+          new Error(`Country provided is not supported, received: ${country}`),
+        );
+      },
+    );
+  });
+
+  describe("shortenPublicHoliday helper tests", () => {
+    holidayMocks.forEach((holiday, index) => {
```
Comment (ShaxxKh) [inline src/__tests__/helpers.test.ts]: it think it is kinda overhead, it is enough to test several cases in different `it`

File: src/services/__tests__/public-holidays.service.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      const result = await getListOfPublicHolidays(year, country);
+      expect(result.length).toBeTruthy();
+    });
+
+    it("shouldn't trigger call and throw an exception in case of wrong country", async () => {
+      const year = 2023;
+      const country = "BY";
+
+      try {
+        await getListOfPublicHolidays(year, country);
```
Comment (ShaxxKh) [inline src/services/__tests__/public-holidays.service.integration.test.ts]: `await expect(getListOfPublicHolidays(year, country)).rejects.toThrowError(`Country provided is not supported, received: ${country}`)`

this is a better approach for error handling testing

File: src/services/__tests__/public-holidays.service.test.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    afterEach(() => {
+      jest.clearAllMocks();
+    });
+
+    it("should trigger call with expected params", async () => {
+      const year = 2023;
+      const country = "GB";
+      const axiosGetSpy = jest
+        .spyOn(axios, "get")
+        .mockImplementation(() => Promise.resolve({ data: holidaysMock }));
```
Comment (ShaxxKh) [inline src/services/__tests__/public-holidays.service.test.ts]: this is a wrapper for `mockImplementation` that automatically resolves
```suggestion
        .mockResolvedValue({ data: holidaysMock });
```

File: src/services/__tests__/public-holidays.service.test.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      const genderResponse = await getListOfPublicHolidays(2023, "GB");
+      expect(genderResponse).toEqual(holidaysMock);
+    });
+
+    it("should simulate call and throw exception", async () => {
+      jest
+        .spyOn(axios, "get")
+        .mockImplementation(() => Promise.reject(new Error("some error")));
```
Comment (ShaxxKh) [inline src/services/__tests__/public-holidays.service.test.ts]:
```suggestion
        .mockRejectedValue(new Error("some error"));
```

File: src/services/__tests__/public-holidays.service.test.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      jest.clearAllMocks();
+    });
+
+    it("should trigger call with expected params", async () => {
+      const year = 2023;
+      const country = "GB";
+      const axiosGetSpy = jest
+        .spyOn(axios, "get")
+        .mockImplementation(() => Promise.resolve({ data: holidaysMock }));
+      const genderResponse = await getListOfPublicHolidays(year, country);
```
Comment (ShaxxKh) [inline src/services/__tests__/public-holidays.service.test.ts]: you also need to test the result of function call

**Review by ShaxxKh — COMMENTED** [review]
There are some other places that need to be improved the same way, please go through the tests and apply the improvements everywhere

File: src/__tests__/helpers.test.ts (hunk 2 above)
Comment (Xelsoid) [reply src/__tests__/helpers.test.ts]: fixed

File: src/services/__tests__/public-holidays.service.integration.test.ts (hunk 3 above)
Comment (Xelsoid) [reply src/services/__tests__/public-holidays.service.integration.test.ts]: fixed

File: src/services/__tests__/public-holidays.service.test.ts (hunk 4 above)
Comment (Xelsoid) [reply src/services/__tests__/public-holidays.service.test.ts]: fixed

File: src/services/__tests__/public-holidays.service.test.ts (hunk 5 above)
Comment (Xelsoid) [reply src/services/__tests__/public-holidays.service.test.ts]: fixed

File: src/services/__tests__/public-holidays.service.test.ts (hunk 6 above)
Comment (Xelsoid) [reply src/services/__tests__/public-holidays.service.test.ts]: the result of function is tested in the other place this test is mostly about checking 'URL' it was called,
removed "const genderResponse ="

**Review by ShaxxKh — APPROVED** [review]
Goodb job

**Pull request merged**